### PR TITLE
Add emscripten as a platform

### PR DESCRIPTION
**Pull request by DerThorsten**
### Description

While emscripten-wasm32 has been added as a [subdirectory](https://github.com/conda/conda/blob/e8c2d9fd8606299f86b1b120ef434975f37ff7b1/conda/base/constants.py#L71), it is missing as a platform.
As a consequence packages from emscripten-forge cannot be installed with conda (https://github.com/mamba-org/mamba/issues/3291).

We are only facing this issue recently since we changed to rattler-build. Before that we use boa to build the packages for emscripten-forge. Boa wrongly used "linux" as platform for emscripten-wasm32 packages instead of "emscripten", therefore we did not see this issue before switching to rattler-build.

Thanks @trungleduc for helping in finding the solution to the problem.

### Checklist - did you ...

- [x] Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?
     
- [x] Add / update necessary tests?
- [x] Add / update outdated documentation?

### CODE REVIEW

**Comment by DerThorsten** [comment]
@conda-bot ping

**Comment by travishathaway** [comment]
@conda-bot check

File: conda/models/enums.py
Comment (jezdez) [inline conda/models/enums.py:48]:
```suggestion
    emscripten = "emscripten"
    wasi = "wasi"
```

**Review by jezdez — COMMENTED** [review]
A few things that need to be done:

- the `Arch` enum also needs a `wasm32` addition
- the `Platform` enum also needs a `wasi` platform
- we need to add a news file that mentions that this is a bugfix for #13095

Additionally, I worry that this wasn't caught earlier, could we add some tests?

**Comment by dbast** [comment]
two questions:
* Do the other Arch and Platform enums have tests?
* What is the value of enum unit tests here? 
 
the thing that really matters is: will the env creation commands listed [here](https://emscripten-forge.readthedocs.io/en/latest/src/usage.html) really work after merging this PR or are we still missing something? Only integration tests or a manual test can ensure this now (and for future changes). If we don't want tests that rely on outside services, then the PR is complete? or we write an complicated test that monkey patches the enums with artificial random arch/platform and mocks a channel+platform to prove that extending the arch + platform + subdir is enough for any additions (this would be imho unfair to request from the PR author as this goes beyond fixing PR https://github.com/conda/conda/pull/13095)

**Comment by dbast** [comment]
a canary build of the PR would be nice to install conda locally and run some of the commands here https://emscripten-forge.readthedocs.io/en/latest/src/usage.html

**Comment by DerThorsten** [comment]
> a canary build of the PR would be nice to install conda locally and run some of the commands here https://emscripten-forge.readthedocs.io/en/latest/src/usage.html

FYI the docs moved to https://emscripten-forge.org/

**Comment by DerThorsten** [comment]
>  If we don't want tests that rely on outside services, then the PR is complete? 

So from my side, I believe its complete.

**Review by wolfv — APPROVED** [review]

**Review by SylvainCorlay — APPROVED** [review]

**Comment by SylvainCorlay** [comment]
@jezdez @jaimergp any blocker on your end?

**Review by trungleduc — APPROVED** [review]

File: news/13962-emscripten-platform
Comment (jaimergp) [inline news/13962-emscripten-platform:7]:
```suggestion
* Add missing `emscripten` and `wasi` entries to the recognized platforms, and `wasm32` to the recognized architectures. (#13095)
```

**Review by jaimergp — DISMISSED** [review]
Just a tiny nit for clarity in the news file.

**Review by jaimergp — DISMISSED** [review]

**Comment by jezdez** [comment]
@DerThorsten I've already left a review for this above, we're fixing a buggy new platform here and you correctly identified that the previous PR essentially wasn't enough to support this feature.

We should make sure that within the conda code base this is fully supported and tested accordingly like other platforms. I understand you're just trying to unblock emscripten-forge and think it would suffice to add an integration test, that runs something similar to:

```
conda create \
    --platform=emscripten-32 \
    --yes \
    -c https://repo.mamba.pm/emscripten-forge \
    -c conda-forge
    pyjs
```

I don't think we'd need to additionally run empack on the envionment to test it, but maybe it'd be fully end-to-end that way?

**Comment by jaimergp** [comment]
A tiny edit to that code block to make it more robust and faster (skip defaults if present):

```
conda create \
    --platform=emscripten-32 \
    --prefix={tmp_path} \
    --yes \
    --override-channels \
    -c https://repo.mamba.pm/emscripten-forge \
    -c conda-forge
    pyjs
```

**Comment by DerThorsten** [comment]
I added a test but I am running into problems installing conda / the test environment locally on my apple-silicon machine,..so I did not run this test whatsoever

File: tests/cli/test_cli_install.py
Comment (wolfv) [inline tests/cli/test_cli_install.py:110]: This is wrong, it should be `emscripten-wasm32`, right?

**Comment by wolfv** [comment]
Do you have any of these tests for `zos-z` or `ppc64le`? It seems like a very straightforward thing to add a new platform that should not require any specific tests. A careful review for typos should be enough.

File: tests/cli/test_cli_install.py
Comment (DerThorsten) [reply tests/cli/test_cli_install.py]: correct =)

File: tests/cli/test_cli_install.py
Comment (jaimergp) [inline tests/cli/test_cli_install.py:118]:
```suggestion
    with tmp_env(
        "--platform=emscripten-wasm32",
        "--override-channels",
        "-c",
        "https://repo.mamba.pm/emscripten-forge",
        "-c",
        "conda-forge",
        "pyjs",
     ) as prefix:
        pass
```

I think this is what we need here.

File: tests/cli/test_cli_install.py
Comment (jaimergp) [inline tests/cli/test_cli_install.py:99]:
```suggestion
```

File: tests/cli/test_cli_install.py
Comment (jaimergp) [inline tests/cli/test_cli_install.py:102]:
```suggestion
```

File: tests/cli/test_cli_install.py
Comment (jezdez) [inline tests/cli/test_cli_install.py:102]: Couldn't that be done with a `--yes`?

**Comment by jezdez** [comment]
pre-commit.ci autofix

**Review by jezdez — DISMISSED** [review]

**Comment by jezdez** [comment]
@kenodegard Do you have an idea why we're not seeing the error output from `subprocess_call_with_clean_env` in the failing test?

**Comment by kenodegard** [comment]
@jezdez when an error or non-zero exit occurs the stdout & stderr are formatted and logged to `INFO`, unfortunately the default logging in conda is `WARNING` so the messages are silenced

**Comment by kenodegard** [comment]
Hm [the test failure](https://github.com/conda/conda/actions/runs/9959794026/job/27517500395?pr=13962#step:10:911) is complaining about emscripten not being a valid platform

**Comment by jaimergp** [comment]
I think the pkgs cache (which is shared across CI runs) got polluted with the test files, and now the vanilla conda (outside this PR) cannot process its own PackageCacheRecord. Let's see if clearing it up helps.

**Comment by jaimergp** [comment]
I manually cleared `integration-1` caches triggered by `refs/pull/13962/merge` from https://github.com/conda/conda/actions/caches, and added a temporary `pkgs_dirs` workaround so they don't get cached again. Hopefully that's enough.

File: tests/cli/test_cli_install.py
Comment (kenodegard) [inline tests/cli/test_cli_install.py:115]: I see you removed the mocker patch but we still need to add `--yes`:

```suggestion
        "--channel=https://repo.mamba.pm/emscripten-forge",
        "--channel=conda-forge",
        "pyjs",
        "--yes",
```

**Review by kenodegard — APPROVED** [review]

**Comment by wolfv** [comment]
yay! 🎉

**Pull request merged**
